Alpaca-Finance-Agents/GuardAlert/src/agent.spec.ts

```diff
@@ -0,0 +1,154 @@
+import { HandleBlock } from "forta-agent";
+import { provideHandleBlock } from "./agent";
+import { createFinding, AddressesRegistry, PriceStatus } from "./utils";
+import { when, resetAllWhenMocks } from "jest-when";
+import {
+  createAddress,
+  encodeParameter,
+  TestBlockEvent,
+} from "forta-agent-tools";
+import { isCallToIsStable } from "./mock.utils";
+
+const fetcherMock = jest.fn();
+const callMock = jest.fn();
+const ethersMock = {
+  call: callMock,
+  _isProvider: true, // This is necessary for being an ethers provider
+} as any;
+
+const workers = [
+  createAddress("0x3"),
+  createAddress("0x4"),
+  createAddress("0x5"),
+];
+
+const lpTokens = [createAddress("0x6"), createAddress("0x7")];
+
+const testAddresses: AddressesRegistry = {
+  vaults: [
+    {
+      address: createAddress("0x1"),
+      workers: [
+        {
+          address: workers[0],
+          lpToken: lpTokens[0],
+        },
+        {
+          address: workers[1],
+          lpToken: lpTokens[0],
+        },
+      ],
+    },
+    {
+      address: createAddress("0x2"),
+      workers: [
+        {
+          address: workers[2],
+          lpToken: lpTokens[1],
+        },
+      ],
+    },
+  ],
+  workerConfig: createAddress("0x0"),
+};
+
+describe("Alpaca Guard Test Suite", () => {
+  let handleBlock: HandleBlock;
+
+  beforeAll(() => {
+    handleBlock = provideHandleBlock(fetcherMock, "testURL", ethersMock);
+  });
+
+  beforeEach(() => {
+    resetAllWhenMocks();
+    when(fetcherMock)
+      .calledWith(expect.anything())
+      .mockReturnValue(testAddresses);
+  });
+
+  it("should return empty findings if not price is deviated", async () => {
+    when(callMock)
+      .calledWith(isCallToIsStable(workers[0]), undefined)
+      .mockReturnValue(encodeParameter("bool", true));
+    when(callMock)
+      .calledWith(isCallToIsStable(workers[2]), undefined)
+      .mockReturnValue(encodeParameter("bool", true));
+
+    const blockEvent = new TestBlockEvent();
+
+    const findings = await handleBlock(blockEvent);
+
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should return a finding if a price is deviated up", async () => {
+    const encodedError = encodeParameter(
+      "string",
+      "WorkerConfig::isStable:: price too high"
+    );
+
+    when(callMock)
+      .calledWith(isCallToIsStable(workers[0]), undefined)
+      .mockReturnValue(encodeParameter("bool", true));
+    when(callMock)
+      .calledWith(isCallToIsStable(workers[2]), undefined)
+      .mockReturnValue("0x08c379a0" + encodedError.slice(2));
+
+    const blockEvent = new TestBlockEvent();
+
+    const findings = await handleBlock(blockEvent);
+
+    expect(findings).toStrictEqual([
+      createFinding(lpTokens[1], PriceStatus.HIGH),
+    ]);
```

Comment:
More for my own understanding, although you don't call any additional methods on `TestBlockEvent()`, because `ethersMock` has `callMock` as the value on one of its keys, that is how it is passed to `provideHandleBlock()`, right? Then that's where `findings` gets its values, right? Because `callMock` saves its state when called in L90 and L93. Am I understanding correctly?